### ISSUE TITLE
Using data in type should be bound to correct child property same as this.

### ISSUE THREAD

**Issue by matjaz** [firebase/bolt#33]
```typescript
type IncrementalTimestamp extends Number {
  validate() { return this > data; }
}

type Test {
  ts: IncrementalTimestamp
}

path /test is Test {}
```

Current output
`newData.child('ts').isNumber() && newData.child('ts').val() > data.val()`

Expected
 `newData.child('ts').isNumber() && newData.child('ts').val() > data.child('ts').val()`

**Comment by mckoss** [comment]
Good catch!

**Comment by matjaz** [comment]
I think there is also a valid use case for `root` in type.

```typescript
type UserId extends String {
  validate() {
    // user exists
    return root.users[this] != null;
  }
}
```

**Comment by mckoss** [comment]
Fixed by 3a5b4bb3d7e6b357e71a3b8f4c77ad16591b4686
